lib/HdtDocument.cc

```diff
@@ -368,6 +369,111 @@ NAN_METHOD(HdtDocument::SearchTerms) {
+  HdtDocument* document;
+  // JavaScript function arguments
+  string filterVal;
+  hdt::TripleComponentRole filterPos;
+  uint32_t limit;
+  hdt::TripleComponentRole position;
+  Persistent<Object> self;
+  // Callback return values
+  vector<string> distinctTerms;
+public:
+  FetchDistinctTermsWorker(HdtDocument* document, char* filterVal, uint32_t filterPos,
+                    uint32_t limit, uint32_t posId, Nan::Callback* callback,
+                    Local<Object> self)
+    : Nan::AsyncWorker(callback),
+      document(document), filterVal(filterVal), filterPos((TripleComponentRole)  filterPos),
+      limit(limit), position((TripleComponentRole) posId) {
+        SaveToPersistent("self", self);
+      };
+
+  void Execute() {
+
+    const char *subject = "";
+    const char *predicate = "";
+    const char *object = "";
+
+    // Initialize filter search term.
+    switch (filterPos) {
+      case SUBJECT:
+        subject = filterVal.c_str();
+        break;
+      case PREDICATE:
+        predicate = filterVal.c_str();
+        break;
+      case OBJECT:
+        object = filterVal.c_str();
+        break;
+    };
+
+    hdt::IteratorUCharString *it = NULL;
+    try {
+
+      Dictionary* dict = document->GetHDT()->getDictionary();
+      it = dict->getPredicates();
+
+      // Iterate over all predicates.
+      while (it->hasNext()) {
+        unsigned char *s = it->next();
+        predicate = reinterpret_cast<char*>(s);
+
+        // Check occurrence.
+        hdt::IteratorTripleString *rit = document->GetHDT()->search(subject, predicate, object);
+        if(rit->hasNext()) {
+          distinctTerms.push_back(predicate);
+        }
+
+        // Exit if reached limit.
+        if(limit && distinctTerms.size() == limit){
+          delete rit;
+          it->freeStr(s);
+          break;
+        }
+
+        delete rit;
+        it->freeStr(s);
+      }
+    }
+    catch (const runtime_error error) { SetErrorMessage(error.what()); }
+    if (it)
+      delete it;
+  }
+
+  void HandleOKCallback() {
+    Nan::HandleScope scope;
+    // Convert the distinctTerms into a JavaScript array
+    uint32_t count = 0;
+    Local<Array> distinctTermsArray = Nan::New<Array>(distinctTerms.size());
+    for (vector<string>::const_iterator it = distinctTerms.begin(); it != distinctTerms.end(); it++)
+      Nan::Set(distinctTermsArray, count++, Nan::New(*it).ToLocalChecked());
+
+    // Send the JavaScript array through the callback
+    const unsigned argc = 2;
+    Local<Value> argv[argc] = { Nan::Null(), distinctTermsArray};
+    callback->Call(GetFromPersistent("self")->ToObject(), argc, argv);
+  }
+
+  void HandleErrorCallback() {
+    Nan::HandleScope scope;
+    Local<Value> argv[] = { Exception::Error(Nan::New(ErrorMessage()).ToLocalChecked()) };
+    callback->Call(GetFromPersistent("self")->ToObject(), 1, argv);
+  }
+};
+
+// Fetches distinct list of predicates given an object.
```

Comment:
…but the code itself seems to take subject/predicate/object. How come?
Also, what would be the utility of passing predicate?